packages/server-wallet/src/wallet/wallet.ts

```diff
@@ -321,6 +331,137 @@ export class Wallet extends EventEmitter<WalletEvents> {
     });
   }
 
+  private emitObjectiveEvents(result: MultipleChannelOutput | SingleChannelOutput): void {
+    if (isMultipleChannelOutput(result)) {
+      // Receiving messages from other participants may have resulted in completed objectives
+      for (const o of result.completedObjectives) {
+        this.emit('ObjectiveCompleted', o);
+      }
+
+      // Receiving messages from other participants may have resulted in new proposed objectives
+      for (const o of result.newObjectives) {
+        if (o.status === 'pending') {
+          this.emit('ObjectiveProposed', o);
+        }
+      }
+    } else {
+      if (hasNewObjective(result)) {
+        if (result.newObjective.status === 'pending') {
+          this.emit('ObjectiveProposed', result.newObjective);
+        }
+      }
+    }
+  }
+
+  /**
+   * Waits for the transactions.wait() promise to resolve on all transaction responses.
+   * @param response
+   */
+  private async waitForTransactions(
+    response: Promise<providers.TransactionResponse[]>
+  ): Promise<void> {
+    const transactions = await response;
+    await Promise.all(transactions.map(tr => tr.wait()));
+  }
+
+  /**
+   * Emits events, sends messages and requests transactions based on the output of the engine.
+   * @param output
+   */
+  private async handleEngineOutput(
+    output: MultipleChannelOutput | SingleChannelOutput
+  ): Promise<void> {
+    this.emitObjectiveEvents(output);
+    await this._messageService.send(getMessages(output));
+    await this.waitForTransactions(this._chainService.handleChainRequests(output.chainRequests));
+  }
+
+  /**
+   * Creates a listener that will be subscribed to the chain service.
+   * It creates a listener for a specific channel id.
+   * @param channelIdToListenFor The channel Id the listener is for
+   * @returns A chain event subscriber that can be passed into the chain service.
+   */
+  private createChainListener(channelIdToListenFor: string): ChainEventSubscriberInterface {
```

Comment:
nit: it would have been nice to move the chain listener construction to the wallet in its own commit, to facilitate reviewing. As is, it is difficult to see which functionality is new, and what is copy/pasted.